### PR TITLE
storage: continually advance upper of migrated builtin tables to track txns upper

### PR DESCRIPTION
**Pull request by aljoscha**
Work towards zero-downtime upgrades: #27406

This makes it so that shards of migrated builtin tables remain
readable at a recent timestamp. We have to treat them separately because
we cannot yet register them in txns.

Importantly, we are careful not to advance the upper further than the
upper of the txn shard because those tables will have to be registered
in txns once a new deployment restarts in read-write mode.

I recommend reviewing this commit by commit!

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by aljoscha** [comment]
fyi @benesch

File: src/storage-controller/src/persist_handles/read_only_table_worker.rs [hunk 1]
```diff
@@ -47,25 +48,59 @@ pub(crate) async fn read_only_mode_table_worker<
...
-        }
+    let gen_upper_future = |mut handle: WriteHandle<SourceData, (), T, i64>| {
+        let fut = async move {
+            let current_upper = handle.shared_upper();
+            handle.wait_for_upper_past(&current_upper).await;
+            let new_upper = handle.shared_upper();
+            (handle, new_upper)
+        };
+
+        fut
```
Comment (jkosh44) [inline src/storage-controller/src/persist_handles/read_only_table_worker.rs:64]: could you just call `.boxed()` here?

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/storage-controller/src/lib.rs [hunk 2]
```diff
@@ -840,18 +840,33 @@ where
...
+                //
+                // We only want to register migrated tables, though, and leave
+                // existing tables out/never write to them in read-only mode.
+                table_registers
+                    .retain(|(id, _write_handle)| migrated_storage_collections.contains(id));
+
+                self.persist_table_worker
+                    .register(register_ts, table_registers)
+                    .await
+                    .expect("table worker unexpectedly shut down");
```
Comment (benesch) [inline src/storage-controller/src/lib.rs:863]: It seems mildly weird to duplicate the call to `register` when it's identical in both blocks, but I'm assuming you wrote it this way on purpose? Maybe to make it more clear how to evolve this code when we rip out the current approach to built-in migrations?

File: src/storage-controller/src/persist_handles/read_only_table_worker.rs [hunk 3]
```diff
@@ -249,6 +252,10 @@ async fn advance_uppers<T>(
     let mut all_updates = BTreeMap::default();
 
     for (id, write_handle) in write_handles.iter_mut() {
+        // As of today, we can only migrate builtin (system) tables, and so only
+        // expect to register those in this read-only table worker.
+        assert!(id.is_system(), "trying to register non-system id {id}");
```
Comment (benesch) [inline src/storage-controller/src/persist_handles/read_only_table_worker.rs:257]: Smart thinking.

**Review by benesch — APPROVED** [review]
As slightly gross as this is, I think we're making the right tradeoffs here. Glad my intuition here ended up lining with your and Joe's Heavy Thinking. Thanks for writing this all up. It looks really good to me, and the refactors (and commit splits!) are very much appreciated.

**Comment by benesch** [comment]
I'm going to merge this so that I can work on https://github.com/MaterializeInc/materialize/pull/28920 without conflicting!

**Pull request merged**